### ISSUE TITLE
Question of tracking the untrackable

### ISSUE THREAD

**Issue by litingfeng** [abhineet123/Deep-Learning-for-Tracking-and-Detection#1]
Hi,

I just read the paper _Tracking The Untrackable: Learning to Track Multiple Cues with Long-Term Dependencies_ and I have a question which I hope you can give me some hints:

What is the dimension of the similarity score ( vector or number)? Say if t_i connects d_j, is the score denoted as [<a href="https://www.codecogs.com/eqnedit.php?latex=\phi(t_i,d_j)" target="_blank"><img src="https://latex.codecogs.com/gif.latex?\phi(t_i,d_j)" title="\phi(t_i,d_j)" /></a>](url) or is the output of some process of [<a href="https://www.codecogs.com/eqnedit.php?latex=\phi(t_i,d_j)" target="_blank"><img src="https://latex.codecogs.com/gif.latex?\phi(t_i,d_j)" title="\phi(t_i,d_j)" /></a>](url)?

I am looking forward to your answer. Thank you very much,

**Comment by abhineet123** [comment]
The similarity score is a single number that is produced by the target RNN using the feature vector <a href="https://www.codecogs.com/eqnedit.php?latex=\phi(t_i,d_j)" target="_blank"><img src="https://latex.codecogs.com/gif.latex?\phi(t_i,d_j)" title="\phi(t_i,d_j)" /></a> as input.

**Comment by litingfeng** [comment]
@abhineet123 Did you mean that [<a href="https://www.codecogs.com/eqnedit.php?latex=\phi(t_i,d_j)" target="_blank"><img src="https://latex.codecogs.com/gif.latex?\phi(t_i,d_j)" title="\phi(t_i,d_j)" /></a>](url) is the input of target RNN(O) and score is the output? But in Figure 2. , [<a href="https://www.codecogs.com/eqnedit.php?latex=\phi(t_i,d_j)" target="_blank"><img src="https://latex.codecogs.com/gif.latex?\phi(t_i,d_j)" title="\phi(t_i,d_j)" /></a>](url) is the output of fc layer following RNN(O).

**Comment by abhineet123** [comment]
Yes, that is correct.
The target RNN applies the softmax classifier and cross entropy loss to the feature vector <a href="https://www.codecogs.com/eqnedit.php?latex=\phi(t_i,d_j)" target="_blank"><img src="https://latex.codecogs.com/gif.latex?\phi(t_i,d_j)" title="\phi(t_i,d_j)" /></a> to produce the similarity score.

This is mentioned in the last line of 3.5 (ii) (first para of column 2 on page 5):
> 
"Our target RNN is also trained to perform the task of data association – outputs the score of whether a detection (d) corresponds to a target (t) from <a href="https://www.codecogs.com/eqnedit.php?latex=\phi(t_i,d_j)" target="_blank"><img src="https://latex.codecogs.com/gif.latex?\phi(t_i,d_j)" title="\phi(t_i,d_j)" /></a> using a Softmax classiﬁer and cross-entropy loss."

It seems that the target RNN produces both <a href="https://www.codecogs.com/eqnedit.php?latex=\phi(t_i,d_j)" target="_blank"><img src="https://latex.codecogs.com/gif.latex?\phi(t_i,d_j)" title="\phi(t_i,d_j)" /></a> and the similarity score.

**Comment by litingfeng** [comment]
@abhineet123 I think I can understand it now. Thank you very much for your elaborate explanation.

**Comment by litingfeng** [comment]
Hi, 

I have another question here about Figure 3. In the second paragraph on page 4( section 3.2), [<a href="https://www.codecogs.com/eqnedit.php?latex=\phi_{t}^{A}" target="_blank"><img src="https://latex.codecogs.com/gif.latex?\phi_{t}^{A}" title="\phi_{t}^{A}" /></a>](url) is a 500-D vector, [<a href="https://www.codecogs.com/eqnedit.php?latex=\phi_{j}" target="_blank"><img src="https://latex.codecogs.com/gif.latex?\phi_{j}" title="\phi_{j}" /></a>](url) is a H-D vector, which is 128 according to **Implementation Details**(2nd paragraph page 6). However, both of them are the output of the same CNN, why is it different of their dimension?

**Comment by abhineet123** [comment]
The paper does not clearly mention how the same CNN is outputting both 500 and 128-D feature vectors but Fig. 3 does show what looks like an extra layer on top of the CNNs corresponding to the 500-D outputs which might indicate another layer that performs this conversion.
This seems to be confirmed by the last para of 3.2 that mentions that they used a pre-trained VGGNet as the appearance feature extractor after replacing its last FC layer with one of their own that produces 500-D vectors.

The reason behind this difference seems simple enough.
The 500-D vectors <a href="https://www.codecogs.com/eqnedit.php?latex=\phi_{t}^{A}" target="_blank"><img src="https://latex.codecogs.com/gif.latex?\phi_{t}^{A}" title="\phi_{t}^{A}" /></a> correspond to the appearance history of the target and are all passed through the LSTM to generate a single H dimensional vector <a href="https://www.codecogs.com/eqnedit.php?latex=\phi_{i}" target="_blank"><img src="https://latex.codecogs.com/gif.latex?\phi_{i}" title="\phi_{i}" /></a> that represents the overall target appearance by fusing information from all of these vectors.

This vector is directly comparable to the H-D feature vector <a href="https://www.codecogs.com/eqnedit.php?latex=\phi_{j}" target="_blank"><img src="https://latex.codecogs.com/gif.latex?\phi_{j}" title="\phi_{j}" /></a> corresponding to the candidate detection which is probably produced by the CNN without this last FC layer that they added.
The two vectors are thus concatenated to generate the 2H-D vector that is finally processed by the Siamese classification network whose output is also a 500-D vector.
Sec. 4.3 mentions that the Siamese classification network is constructed using the same CNN as used for appearance feature extraction.
This suggests that the FC layer of the Siamese network that produces this final 500-D output is similar to the FC layer that converts the H-D CNN output to the 500-D vectors <a href="https://www.codecogs.com/eqnedit.php?latex=\phi_{t}^{A}" target="_blank"><img src="https://latex.codecogs.com/gif.latex?\phi_{t}^{A}" title="\phi_{t}^{A}" /></a>. Since it is a Siamese network, however, it contains a pair of these feature extractor CNNs and its FC layer has been trained to distinguish between _two_ of these H-D vectors instead of simply mapping one into a 500-D vector.

**Comment by litingfeng** [comment]
In summary, you mean that the 500-D vector of [<a href="https://www.codecogs.com/eqnedit.php?latex=\phi_{t}^{A}" target="_blank"><img src="https://latex.codecogs.com/gif.latex?\phi_{t}^{A}" title="\phi_{t}^{A}" /></a>](url) is the output of the feature extractor CNN with a 500 units of fc layer following a 128 units of fc layer, and the [<a href="https://www.codecogs.com/eqnedit.php?latex=\phi_{j}" target="_blank"><img src="https://latex.codecogs.com/gif.latex?\phi_{j}" title="\phi_{j}" /></a>](url) is the output of incomplete version of CNN which doesn't have the 500 units fc layer. I'm not sure whether I comprehend your point in the right way. In addition, I don't quite understand the last sentence and the architecture of the siamese network. Does it look like the figure ? 
![1](https://user-images.githubusercontent.com/13741167/31710319-35f03e68-b427-11e7-88f7-219566222588.png)

**Comment by abhineet123** [comment]
Yes that is what I mean.
It is impossible to say exactly how the Siamese network is designed until they release their code but yes, this is roughly what I had in mind.

**Comment by litingfeng** [comment]
@abhineet123 I am really grateful for your reply. Later I will ask the author for more details.

**Comment by abhineet123** [comment]
Glad to be of assistance and please let me know what the authors have to say about this.

**Comment by litingfeng** [comment]
Hi,

I asked the author but haven't received response yet. Here I have another question： do you know how to train the LSTM in this appearance model？ Is LSTM included in siamese CNN？ I'm still connused about the training procedure. Thank you very much.

**Comment by abhineet123** [comment]
No I am also waiting for the authors to release their code to get the details of the training procedure.

**Comment by swamika001** [comment]
Hi,
In **3.3 Motion**, the authors wrote that velocities are extracted by their motion feature extractor. Does anyone have a clue on what algorithm it could be ?

**Comment by abhineet123** [comment]
Probably some kind of optical flow algorithm like [cvCalcOpticalFlowPyrLK](https://docs.opencv.org/2.4/modules/video/doc/motion_analysis_and_object_tracking.html#calcopticalflowpyrlk).
This is what they used in an [earlier version](https://github.com/abhineet123/Deep-Learning-for-Tracking-and-Detection/blob/master/Tracking/RL/Learning%20to%20Track%20Online%20Multi-object%20Tracking%20by%20Decision%20Making%20%20iccv15.pdf) of this paper.

**Comment by litingfeng** [comment]
I just re-read the paper and have another question:

Why is the dimension of [<a href="https://www.codecogs.com/eqnedit.php?latex=\phi_{A}" target="_blank"><img src="https://latex.codecogs.com/gif.latex?\phi_{A}" title="\phi_{A}" /></a>](url) set to 500? 

Since the vector encodes whether a box at time _t+1_ corresponds to a specific target, the dimension should be the same as the number of targets. After softmax, each unit encodes the score of the similarity score between the detection and the target. 500 is too large to be the number of targets.

**Comment by nidhinkrishnanv** [comment]
Hi,

In 3.5 Target, any idea what is the input sequence to the Target RNN. The authors mention that the output of appearance, motion and interaction are concatenated and passed to Target RNN. But then how do that result in a sequence?

**Comment by tonmoyborah** [comment]
Has the code been released yet or any implementation available?

**Comment by abhineet123** [comment]
Not as far as I know.

**Comment by behappyZheng** [comment]
hi, is the the paper Tracking The Untrackable: Learning to Track Multiple Cues with Long-Term Dependencies have code to be implementation

**Comment by abhineet123** [comment]
Not that I am aware of.

**Comment by icesohelrana** [comment]
Do you have any intuition about input image size fed into CNN? VGG16 takes 224*224 size as input and produces 28055 as first FC layer. But, a person's size(height and width) would not be a square. So how they cropped the image? If input image size is different, then first FC layer will be different.

**Comment by abhineet123** [comment]
In their [earlier paper](https://github.com/abhineet123/Deep-Learning-for-Tracking-and-Detection/blob/master/Tracking/RL/Learning%20to%20Track%20Online%20Multi-object%20Tracking%20by%20Decision%20Making%20%20iccv15.pdf), they extract the patch and then resize it to a fixed size (224*224 in your case) without preserving the aspect ratio. Though the patch becomes distorted to human eyes, it probably doesn't make any difference to the CNN as long as test patches are distorted in the same way as training ones.

**Comment by borguleabhijeet** [comment]
I'm trying to the experiment after long but it's not working for me.
I tried the following but little different way:

Ex.

Track1->BB1,BB2,BB3,BB4,BB5,BB6,               D0(Detection)

BB1,BB2,BB3,BB4,BB5,BB6->VGG->LSTM=====> (512+512+...+512)
              shared weights                                                                      Concat(512+3072) FC (512)  Softmax(2)
 D0->VGG->======================================>(512)

Do you know any other way? or did you find any code?

**Comment by Archie2035** [comment]
hi, is the the paper Tracking The Untrackable: Learning to  Track Multiple Cues with Long-Term Dependencies have code to be  implementation?

**Comment by abhineet123** [comment]
Not that I'm aware of.

**Comment by Archie2035** [comment]
Is there a tracking method using self-attention（such as transformer、BERT） recently?
